evmbin/src/info.rs

```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
 	})
 }
 
+/// Input data to run transaction
+#[derive(Debug)]
+pub struct TxInput<'a, T> {
+	/// Chain specification name associated with the transaction.
+	pub name: &'a String,
+	/// Transaction index from list of transactions within a state root hashes corresponding to a chain.
+	pub idx: usize,
```

Comment:
Do we really need all the members to be `pub`? It is generally avoided so that the code is easier to change in the future. Maybe it could be a way to use the `Default` trait to avoid supplying all args all the time? It would also let you do fancy things like have `MyChainTxInput` types where e.g. the `name` member is pre-set and you can have logic to ensure users don't misuse the api (say, use a `ForkSpec` that is invalid for the given chain spec name).

Reply:
I didn't figure out how to use the `Default` trait, but I've pushed an approach here  
in this commit https://github.com/paritytech/parity-ethereum/commit/426a4388b0b7989ff2903fe1a07c98ee86642563, but unfortunately I can't work out how to overcome these errors 
<details><p>
error: expected expression, found `}`
   --> evmbin/./src/info.rs:139:3
    |
139 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:146:3
    |
146 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:153:3
    |
153 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:160:3
    |
160 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:167:3
    |
167 |         }
    |         ^ expected expression

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:221:7
    |
221 |                         tx_input::std_json_err_only();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:225:7
    |
225 |                         tx_input::std_json_out_only();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:229:7
    |
229 |                         tx_input::std_json_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:237:7
    |
237 |                         tx_input::json_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:241:7
    |
241 |                         tx_input::simple_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:137:15
    |
137 |             informant: display::std_json::Informant::err_only(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stderr, std::io::Stderr>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:136:3
    |
136 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:144:15
    |
144 |             informant: display::std_json::Informant::out_only(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stdout, std::io::Stdout>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:143:3
    |
143 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:151:15
    |
151 |             informant: display::std_json::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stderr, std::io::Stdout>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:150:3
    |
150 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:158:15
    |
158 |             informant: display::json::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::json::Informant`
    |
    = note: expected type `T`
               found type `display::json::Informant`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:157:3
    |
157 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:165:15
    |
165 |             informant: display::simple::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::simple::Informant`
    |
    = note: expected type `T`
               found type `display::simple::Informant`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:164:3
    |
164 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error: aborting due to 20 previous errors
</p></details>